crates/core/src/ring/mod.rs

```diff
@@ -254,9 +254,43 @@ impl Ring {
         skip_list: impl Contains<PeerId>,
     ) -> Option<PeerKeyLocation> {
         let router = self.router.read();
+        let target_location = Location::from(contract_key);
 
-        self.connection_manager
-            .routing(Location::from(contract_key), None, skip_list, &router)
+        // First check if self should be considered
```

Comment:
I understand the problem, but is important to notice this is a change in the logic with different semantics that we had, which is ok, but we should document here why we are doing this this way for future reference.

The reason we dismissed the self location (specially in puts), is because we want the contract to live in the network and guarantee is not "only" local. I wonder if a better solution is to both putting it into self location (if is a better location than any other connected peers) and the best next peer. So puts effectively would be always double puts (as long as there are connections available, ofc otherwise it will be only in self, but that would only b epossible if u are a gateway putting contracts that shouldn't be something usual).